src/ansi-c/c_typecheck_code.cpp

```diff
@@ -809,6 +818,41 @@ void c_typecheck_baset::typecheck_dowhile(code_dowhilet &code)
 
   typecheck_spec_loop_invariant(code);
   typecheck_spec_decreases(code);
+
+  if(code.find(ID_C_spec_assigns).is_not_nil())
+    typecheck_spec_assigns(
+      static_cast<unary_exprt &>(code.add(ID_C_spec_assigns)).op().operands());
+}
+
+void c_typecheck_baset::typecheck_spec_assigns(exprt::operandst &targets)
+{
+  for(auto &target : targets)
+  {
+    typecheck_expr(target);
+
+    if(target.type().id() == ID_empty)
+    {
+      error().source_location = target.source_location();
+      error() << "void-typed targets not permitted" << eom;
+      throw 0;
+    }
+    else if(target.id() == ID_pointer_object)
```

Comment:
should we error-out on `void *` or any nested `(...((void *)*) ...)` ?

Reply:
This case is hit when the user writes `__CPROVER_POINTER_OBJECT(ptr)`. Yes `ptr` in this case can be `void*` but should we not allow users to declare the underlying object as assignable? May be they are casting it correctly before actually assigning?